components/support/rc_crypto/nss/src/pbkdf2.rs

```diff
@@ -0,0 +1,71 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::util::{ensure_nss_initialized, map_nss_secstatus, sec_item_as_slice, ScopedPtr};
+use crate::{
+    error::*,
+    pk11::{slot::get_internal_slot, sym_key::pbe_key_gen, types::AlgorithmID},
+};
+
+// Expose for consumers to choose the hasing algorithm
+// Currently only SHA256 supported
+pub use crate::pk11::context::HashAlgorithm;
+use nss_sys::SECOidTag;
+use std::convert::TryFrom;
+
+// ***** BASED ON THE FOLLOWING IMPLEMENTATION *****
+// https://github.com/sqlcipher/sqlcipher/blob/4a81bea61e1da6fec222d713852830f1fd01aed2/src/crypto_nss.c#L182-L232
+// Which I believe is based on how it is done on WebCrypto:
+// https://searchfox.org/mozilla-central/rev/8ccea36c4fb09412609fb738c722830d7098602b/dom/crypto/WebCryptoTask.cpp#2567
+
+pub fn pbkdf2_key_derive(
+    password: &mut [u8],
+    salt: &mut [u8],
+    iterations: u32,
+    hash_algorithm: HashAlgorithm,
+    out: &mut [u8],
+) -> Result<()> {
+    ensure_nss_initialized();
+    let oid_tag = match hash_algorithm {
+        HashAlgorithm::SHA256 => SECOidTag::SEC_OID_HMAC_SHA256 as u32,
+    };
+
```

Comment:
Very very nit: there's a ton of line breaks in this function which makes it bit harder to reason about. Can we group instructions somehow? 1. create the alg, 2. stretch the key 3. get the vector (or any other grouping that you like really :))